Sources/Layout/MessageCellLayout.swift

```diff
@@ -63,23 +59,352 @@ public struct TextMessageLayout: MessageCellLayout {
+    }
+
+    internal override func messageContainerMaxWidth(for message: MessageType) -> CGFloat {
+        let maxWidth = super.messageContainerMaxWidth(for: message)
+        let textInsets = messageLabelInsets(for: message)
+        return maxWidth - textInsets.horizontal
+    }
+
+    open override func messageContainerSize(for message: MessageType) -> CGSize {
+        let maxWidth = messageContainerMaxWidth(for: message)
+
+        var messageContainerSize: CGSize
+        let attributedText: NSAttributedString
+
+        switch message.data {
+        case .text(let text):
+            attributedText = NSAttributedString(string: text, attributes: [.font: messageLabelFont])
+        case .attributedText(let text):
+            attributedText = text
+        case .emoji(let text):
+            attributedText = NSAttributedString(string: text, attributes: [.font: emojiLabelFont])
+        default:
+            fatalError("messageContainerSize received unhandled MessageDataType: \(message.data)")
+        }
+
+        messageContainerSize = labelSize(for: attributedText, considering: maxWidth)
+
+        let messageInsets = messageLabelInsets(for: message)
+        messageContainerSize.width += messageInsets.horizontal
+        messageContainerSize.height += messageInsets.vertical
+
+        return messageContainerSize
+    }
 
+    open override func configure(attributes: UICollectionViewLayoutAttributes) {
+        super.configure(attributes: attributes)
+        guard let attributes = attributes as? MessagesCollectionViewLayoutAttributes else { return }
+
+        let dataSource = messagesLayout.messagesDataSource
+        let indexPath = attributes.indexPath
+        let message = dataSource.messageForItem(at: indexPath, in: messagesLayout.messagesCollectionView)
+
+        attributes.messageLabelInsets = messageLabelInsets(for: message)
+        attributes.messageLabelFont = messageLabelFont
+
+        switch message.data {
+        case .attributedText(let text):
+            guard !text.string.isEmpty else { return }
+            guard let font = text.attribute(.font, at: 0, effectiveRange: nil) as? UIFont else { return }
+            attributes.messageLabelFont = font
+        case .emoji:
+            attributes.messageLabelFont = emojiLabelFont
+        default:
+            break
+        }
+    }
+}
+
+open class MediaMessageSizeCalculator: MessageSizeCalculator {
+
+    // TODO: - Figure out how we are going to size these
+    public var incomingMediaSize: CGSize = .zero
+    public var outgoingMediaSize: CGSize = .zero
+
+    open override func messageContainerSize(for message: MessageType) -> CGSize {
+        let maxWidth = messageContainerMaxWidth(for: message)
+        switch message.data {
+        case .photo(_):
+            // TODO: - Implement actual sizing
+            return CGSize(width: maxWidth, height: 240)
+        case .video(_, _):
```

Comment:
I need to figure out how we're going to handle sizing `.photo` and `.video`. I think JSQMVC uses static sizes? Need to check

Reply:
It uses the `JSQMessageMediaData` protocol to handle this.
Also we still need to determine wether we need the hash key to be the key of the cache.

Reply:
Another issue, we still only support one cell to one section? Why we not need one section to all cells if it is?

Reply:
> Another issue, we still only support one cell to one section? What's the point we not choose one section to all cells if it is?

I think we can remove this limitation but I would do this in a later PR. I don't want to make all these changes all at once. I'll still give it a try before merging anything to see if this PR makes that possible or impossible

Reply:
👍